src/base/scene.cu

```diff
@@ -0,0 +1,309 @@
+
+    // Extract normals
+    size_t normalsBufferOffset = normalsBufferView.byteOffset;
+    size_t normalsBufferLength = normalsBufferView.byteLength;
+    // TODO handle byteStride
+    Vector3f *normalsBytes = reinterpret_cast<Vector3f *>(
+        normalsBuffer.data.data() + normalsBufferOffset);
+    std::vector<Vector3f> primitiveNormals(
+        normalsBytes, normalsBytes + normalsBufferLength / sizeof(Vector3f));
+
+    // Extract face indices
+    size_t facesBufferOffset = facesBufferView.byteOffset;
+
+    static const size_t faceCount =
+        facesAccessor.count /
+        TRIANGLE_VERT_COUNT;  // gltf only supports triangles. No quads or
+                              // ngons
+
+    std::vector<Vector3i> primitiveFaces(faceCount);
+
+    // TODO handle byteStride
+
+    if (facesAccessor.componentType == 5123) {  // Component type unsigned short
+      unsigned short *facesBytes = reinterpret_cast<unsigned short *>(
+          facesBuffer.data.data() + facesBufferOffset);
+      for (size_t f = 0; f < faceCount; ++f) {
+        primitiveFaces[f] =
+            Vector3i(facesBytes[f * TRIANGLE_VERT_COUNT] + primitiveOffset,
+                     facesBytes[f * TRIANGLE_VERT_COUNT + 1] + primitiveOffset,
+                     facesBytes[f * TRIANGLE_VERT_COUNT + 2] + primitiveOffset);
+      }
+
+    } else if (facesAccessor.componentType ==
+               5125) {  // Component type unsigned int
+      unsigned int *facesBytes = reinterpret_cast<unsigned int *>(
+          facesBuffer.data.data() + facesBufferOffset);
+      for (size_t f = 0; f < faceCount; ++f) {
+        primitiveFaces[f] =
+            Vector3i(facesBytes[f * TRIANGLE_VERT_COUNT] + primitiveOffset,
+                     facesBytes[f * TRIANGLE_VERT_COUNT + 1] + primitiveOffset,
+                     facesBytes[f * TRIANGLE_VERT_COUNT + 2] + primitiveOffset);
+      }
+    }
+
+    positions.insert(positions.end(), primitivePositions.begin(),
+                     primitivePositions.end());
+    normals.insert(normals.end(), primitiveNormals.begin(),
+                   primitiveNormals.end());
+    faces.insert(faces.end(), primitiveFaces.begin(), primitiveFaces.end());
+
+    primitiveOffset += primitivePositions.size();
+  }
+
+  return std::shared_ptr<Mesh>(new Mesh(positions, normals, faces));
+}
+
+std::shared_ptr<gm::PerspectiveCamera> gm::Scene::loadCamera(
+    const std::vector<int> &node_ids, const tinygltf::Model &model) {
+  // NOTE The following section uses raw pointers to improve performance and
+  // prevent unnecessary copies. Smart pointers can not be used here as we are
+  // not managing the memory to which we are pointing.
+  const tinygltf::Node *camera = nullptr;
+  std::unordered_map<const tinygltf::Node *, const tinygltf::Node *> parents;
+
+  // find camera node
+  for (size_t i = 0; i < node_ids.size() && !camera; ++i) {
+    std::stack<const tinygltf::Node *> toVisit;
```

Comment:
Just curious, why are we doing depth-first over breadth-first search? Is there an advantage of doing one over the other?